pkg/scaling/scale_handler.go

```diff
@@ -681,3 +711,126 @@ func (h *scaleHandler) isScaledJobActive(ctx context.Context, scaledJob *kedav1a
+// what metrics should be used. In case of composite scaler (ComplexScalingLogic struct),
+// all external metrics will be used. Returns all external metrics otherwise it
+// returns the same metric given.
+func (h *scaleHandler) getTrueMetricArray(ctx context.Context, metricName string, so *kedav1alpha1.ScaledObject) ([]string, error) {
+	// if composite scaler is given return all external metrics
+	if so != nil && so.Spec.Advanced != nil && so.Spec.Advanced.ComplexScalingLogic.Target != "" {
+		if len(so.Status.ExternalMetricNames) == 0 {
+			scaledObject := &kedav1alpha1.ScaledObject{}
+			err := h.client.Get(ctx, types.NamespacedName{Name: so.Name, Namespace: so.Namespace}, scaledObject)
+			if err != nil {
+				log.Error(err, "failed to get ScaledObject", "name", so.Name, "namespace", so.Namespace)
+				return nil, err
+			}
+			if len(scaledObject.Status.ExternalMetricNames) == 0 {
+				err := fmt.Errorf("failed to get ScaledObject.Status.ExternalMetricNames, probably invalid ScaledObject cache")
+				log.Error(err, "failed to get ScaledObject.Status.ExternalMetricNames, probably invalid ScaledObject cache", "scaledObject.Name", scaledObject.Name, "scaledObject.Namespace", scaledObject.Namespace)
+				return nil, err
+			}
+
+			so = scaledObject
+		}
+		return so.Status.ExternalMetricNames, nil
+	}
+	return []string{metricName}, nil
+}
+
+// help function to determine whether or not metricName is the correct one.
+// standard function will be array of one element if it matches or none if it doesnt
+// that is given from getTrueMetricArray().
+// In case of compositeScaler, cycle through all external metric names
+func arrayContainsElement(el string, arr []string) bool {
+	for _, item := range arr {
+		if strings.EqualFold(item, el) {
+			return true
+		}
+	}
+	return false
+}
+
+// if given right conditions, try to apply the given custom formula in SO
+func applyComplexLogicFormula(csl kedav1alpha1.ComplexScalingLogic, metrics []external_metrics.ExternalMetricValue, pairList map[string]string) ([]external_metrics.ExternalMetricValue, error) {
+	if csl.Formula != "" {
+		metrics, err := calculateComplexLogicFormula(metrics, csl.Formula, pairList)
+		return metrics, err
+	}
+	return metrics, nil
+}
+
+// calculate custom formula to metrics and return calculated and finalized metric
+func calculateComplexLogicFormula(list []external_metrics.ExternalMetricValue, formula string, pairList map[string]string) ([]external_metrics.ExternalMetricValue, error) {
+	var ret external_metrics.ExternalMetricValue
+	var out float64
+	ret.MetricName = "composite-metric-name"
+	ret.Timestamp = v1.Now()
+
+	// using https://github.com/antonmedv/expr to evaluate formula expression
+	data := make(map[string]float64)
+	for _, v := range list {
+		data[pairList[v.MetricName]] = v.Value.AsApproximateFloat64()
+	}
+	program, err := expr.Compile(formula)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to compile custom formula: %w", err)
+	}
```

Comment:
I guess that this is fast, but in the same way we already caching the scalers to improve the performance, we could cache the compiled expression to improve the performance not needing to compile it on every calculation. 

I think this will be fast, but in high load scenarios this could be a problem (I'm thinking on this as with regex compilations, but maybe I'm wrong)

Reply:
good catch! Agree.